### PR TITLE
Reprocess Addons configuration on temporary error (for example network)

### PR DESCRIPTION
**Pull request by piotrmiskiewicz**
**Description**

Changes proposed in this pull request:
- reprocess (cluster)addonsconfiguration on fetching error

**Related issue(s)**

Related: https://github.tools.sap/kyma/backlog/issues/780

### CODE REVIEW

File: test/integration/suite_test.go [hunk 1]
```diff
@@ -659,7 +674,7 @@ func (ts *testSuite) updateClusterAddonsConfigurationStatusPhase(name string, ph
 func (ts *testSuite) createSpecRepositories(urls []string, repoKind string, repoModifiers ...func(r *v1alpha1.SpecRepository)) []v1alpha1.SpecRepository {
 	// v1alpha1.SpecRepository cannot be null, needs to be empty array
 	if len(urls) == 0 {
-		return []v1alpha1.SpecRepository{{}}
```
Comment (piotrmiskiewicz) [inline test/integration/suite_test.go:662]: it was a big mistake here

**Comment by piotrmiskiewicz** [comment]
How to test:
1. Install Helm broker (see the HB docs)
2. Create ClusterAddonsConfiguration with an url to an existing addons github repo but not existing branch, for example:
3. github.com/piotrmiskiewicz/bundles//addons/index-testing.yaml?ref=not-existing
4. Look at HB logs
5. Create and push a `not-existing` branch, then wait for ClusterAddonsConfiguration to be ready

File: internal/controller/common.go [hunk 2]
```diff
@@ -133,10 +143,42 @@ func (c *common) PrepareForProcessing(addon *internal.CommonAddon) error {
 	return nil
 }
 
+type TemporaryError struct {
```
Comment (jasiu001) [inline internal/controller/common.go:146]: if this is used only in one case, maybe the standard error is enough:
```go
import (
  basicError "errors"
)
(...)
var TemporaryError = basicError.New("temporary error")
(...)
if err := c.OnAdd(addon, lastStatus); err != nil {
  if basicError.Is(err, TemporaryError) {
    c.log.Infof("The error is temporary, marking for a reprocess")
  }
}
  (...)
}
(...)
if repositories.IsRepositoriesFetchingError() {
  c.log.Warnf("Unable to read repositories, scheduling for a retry")
  addon.Status.Phase = v1alpha1.AddonsConfigurationFailed

  c.statusSnapshot(&addon.Status, repositories)
  if !addon.Status.Equals(&lastStatus) {
    if err := c.updateAddonStatus(addon); err != nil {
      return errors.Wrap(err, "while update addons configuration status")
    }
  }
  return TemporaryError
}
```

File: test/integration/integration_test.go [hunk 3]
```diff
@@ -265,6 +265,25 @@ func TestUnregisteringClusterServiceBroker(t *testing.T) {
 	suite.waitForClusterServiceBrokerNotRegistered()
 }
 
+func TestRetryOnNetworkError(t *testing.T) {
```
Comment (jasiu001) [inline test/integration/integration_test.go:268]: I am not sure if this test works properly but when I change the error returned on `OnAdd` method from temporary to regular, the test still passes. Is this desirable behavior?

Comment (piotrmiskiewicz) [reply test/integration/integration_test.go:268]: Returning regular error makes retry without any delay. Removing the `return` (which makes the code like before the PR) makes the test red.

**Review by jasiu001 — DISMISSED** [review]

File: docs/configuration.md [hunk 4]
```diff
@@ -21,3 +21,4 @@ Use the following environment variables to configure the `Controller` container
 | **APP_CLUSTER_SERVICE_BROKER_NAME** | Yes | | Specifies the name of the ClusterServiceBroker resource which registers the Helm Broker in the Service Catalog. |
 | **APP_DEVELOP_MODE** | No | `false` | If set to `true`, you can use unsecured HTTP-based repositories URLs. |
 | **APP_DOCUMENTATION_ENABLED** | No | `false` | If set to `true`, the Helm Broker uploads addons documentation to the [Rafter](https://kyma-project.io/docs/components/headless-cms/). |
+| **APP_REPROCESS_ON_ERROR_DURATION** | No | `5m` | The duration used to reschedule repository, which fails because of a network error. |
```
Comment (klaudiagrz) [inline docs/configuration.md:24]:
```suggestion
| **APP_REPROCESS_ON_ERROR_DURATION** | No | `5m` | Specifies the time after which Helm Broker performs the repository connection retry that has previously failed. |
```

**Review by klaudiagrz — APPROVED** [review]

**Review by jasiu001 — APPROVED** [review]

**Pull request merged**